File: prayer_app_expo/src/services/locationService.js
import * as Location from 'expo-location';
import AsyncStorage from '@react-native-async-storage/async-storage';

const KEYS = {
    lat: 'loc_lat',
    lon: 'loc_lon',
    city: 'loc_city',
    country: 'loc_country',
    timezone: 'loc_timezone',
    source: 'loc_source',
};

export const FALLBACK_LOCATION = {
    lat: 44.4268,
    lon: 26.1025,
    city: 'Bucharest',
    country: 'Romania',
    timezone: 'Europe/Bucharest',
    source: 'default',
};

/** Load previously persisted location, or return fallback. */
export async function loadSavedLocation() {
    try {
        const source = await AsyncStorage.getItem(KEYS.source);
        if (!source) return { ...FALLBACK_LOCATION };
        const lat = parseFloat(await AsyncStorage.getItem(KEYS.lat)) || FALLBACK_LOCATION.lat;
        const lon = parseFloat(await AsyncStorage.getItem(KEYS.lon)) || FALLBACK_LOCATION.lon;
        const city = (await AsyncStorage.getItem(KEYS.city)) || 'Unknown';
        const country = (await AsyncStorage.getItem(KEYS.country)) || '';
        const timezone = (await AsyncStorage.getItem(KEYS.timezone)) || FALLBACK_LOCATION.timezone;
        return { lat, lon, city, country, timezone, source };
    } catch (_) {
        return { ...FALLBACK_LOCATION };
    }
}

/** Request permission, get coords, reverse geocode, persist. */
export async function detectLocation() {
    // 1. Request permission
    const { status } = await Location.requestForegroundPermissionsAsync();
    if (status !== 'granted') {
        await persistLocation(FALLBACK_LOCATION);
        return { ...FALLBACK_LOCATION };
    }

    // 2. Get coordinates
    let coords;
    try {
        const pos = await Location.getCurrentPositionAsync({
            accuracy: Location.Accuracy.Low,
        });
        coords = { lat: pos.coords.latitude, lon: pos.coords.longitude };
    } catch (_) {
        await persistLocation(FALLBACK_LOCATION);
        return { ...FALLBACK_LOCATION };
    }

    // 3. Reverse geocode
    let city = 'Unknown';
    let country = '';
    try {
        const results = await Location.reverseGeocodeAsync({
            latitude: coords.lat,
            longitude: coords.lon,
        });
        if (results.length > 0) {
            const r = results[0];
            city = r.city || r.subregion || r.region || 'Unknown';
            country = r.country || '';
        }
    } catch (_) {
        // Keep city = 'Unknown'
    }

    // 4. Timezone — use device's IANA timezone
    const deviceTz = Intl.DateTimeFormat().resolvedOptions().timeZone || FALLBACK_LOCATION.timezone;

    const data = {
        lat: coords.lat,
        lon: coords.lon,
        city,
        country,
        timezone: deviceTz,
        source: 'gps',
    };

    await persistLocation(data);
    return data;
}

async function persistLocation(d) {
    try {
        await AsyncStorage.setItem(KEYS.lat, String(d.lat));
        await AsyncStorage.setItem(KEYS.lon, String(d.lon));
        await AsyncStorage.setItem(KEYS.city, d.city);
        await AsyncStorage.setItem(KEYS.country, d.country);
        await AsyncStorage.setItem(KEYS.timezone, d.timezone);
        await AsyncStorage.setItem(KEYS.source, d.source);
    } catch (_) { /* ignore */ }
}

File: prayer_app_flutter/lib/src/services/location_service.dart
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';
import 'package:geocoding/geocoding.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Persisted location data.
class LocationData {
  final double lat;
  final double lon;
  final String city;
  final String country;
  final String timezone;
  final String source; // "gps" or "default"

  const LocationData({
    required this.lat,
    required this.lon,
    required this.city,
    required this.country,
    required this.timezone,
    required this.source,
  });

  static const LocationData fallback = LocationData(
    lat: 44.4268,
    lon: 26.1025,
    city: 'Bucharest',
    country: 'Romania',
    timezone: 'Europe/Bucharest',
    source: 'default',
  );
}

class LocationService {
  static const String _kLat = 'loc_lat';
  static const String _kLon = 'loc_lon';
  static const String _kCity = 'loc_city';
  static const String _kCountry = 'loc_country';
  static const String _kTz = 'loc_timezone';
  static const String _kSource = 'loc_source';

  /// Load persisted location, or return fallback.
  Future<LocationData> loadSaved() async {
    final prefs = await SharedPreferences.getInstance();
    final source = prefs.getString(_kSource);
    if (source == null) return LocationData.fallback;
    return LocationData(
      lat: prefs.getDouble(_kLat) ?? LocationData.fallback.lat,
      lon: prefs.getDouble(_kLon) ?? LocationData.fallback.lon,
      city: prefs.getString(_kCity) ?? 'Unknown',
      country: prefs.getString(_kCountry) ?? '',
      timezone: prefs.getString(_kTz) ?? LocationData.fallback.timezone,
      source: source,
    );
  }

  /// Request permission, get coords, reverse geocode, persist.
  /// Returns the detected or fallback LocationData.
  Future<LocationData> detect() async {
    // 1. Check / request permission
    LocationPermission perm = await Geolocator.checkPermission();
    if (perm == LocationPermission.denied) {
      perm = await Geolocator.requestPermission();
    }

    if (perm == LocationPermission.denied ||
        perm == LocationPermission.deniedForever) {
      // Permission denied → save and return fallback
      await _persist(LocationData.fallback);
      return LocationData.fallback;
    }

    // 2. Get coordinates
    Position pos;
    try {
      pos = await Geolocator.getCurrentPosition(
        locationSettings: const LocationSettings(
          accuracy: LocationAccuracy.low,
          timeLimit: Duration(seconds: 10),
        ),
      );
    } catch (_) {
      await _persist(LocationData.fallback);
      return LocationData.fallback;
    }

    // 3. Reverse geocode
    String city = 'Unknown';
    String country = '';
    try {
      final placemarks = await placemarkFromCoordinates(pos.latitude, pos.longitude);
      if (placemarks.isNotEmpty) {
        final p = placemarks.first;
        city = p.locality ?? p.subAdministrativeArea ?? p.administrativeArea ?? 'Unknown';
        country = p.country ?? '';
      }
    } catch (_) {
      // Keep city = 'Unknown'
    }

    // 4. Timezone — detect from device
    final deviceTz = _getDeviceTimezone();

    final data = LocationData(
      lat: pos.latitude,
      lon: pos.longitude,
      city: city,
      country: country,
      timezone: deviceTz,
      source: 'gps',
    );

    await _persist(data);
    return data;
  }

  /// Get device IANA timezone string.
  /// DateTime.timeZoneName may return abbreviations on some platforms,
  /// so we try mapping common ones. Falls back to the abbreviation itself.
  String _getDeviceTimezone() {
    final tz = DateTime.now().timeZoneName;
    // If it already looks like IANA (contains '/'), use it directly
    if (tz.contains('/')) return tz;
    // Common abbreviation → IANA mapping
    const abbrevMap = {
      'PST': 'America/Los_Angeles',
      'PDT': 'America/Los_Angeles',
      'MST': 'America/Denver',
      'MDT': 'America/Denver',
      'CST': 'America/Chicago',
      'CDT': 'America/Chicago',
      'EST': 'America/New_York',
      'EDT': 'America/New_York',
      'GMT': 'Europe/London',
      'BST': 'Europe/London',
      'CET': 'Europe/Berlin',
      'CEST': 'Europe/Berlin',
      'EET': 'Europe/Bucharest',
      'EEST': 'Europe/Bucharest',
      'IST': 'Asia/Kolkata',
      'JST': 'Asia/Tokyo',
      'KST': 'Asia/Seoul',
      'AEST': 'Australia/Sydney',
      'AEDT': 'Australia/Sydney',
      'NZST': 'Pacific/Auckland',
      'NZDT': 'Pacific/Auckland',
      'AST': 'Asia/Riyadh',
      'GST': 'Asia/Dubai',
      'PKT': 'Asia/Karachi',
      'WIB': 'Asia/Jakarta',
      'SGT': 'Asia/Singapore',
      'HKT': 'Asia/Hong_Kong',
      'PHT': 'Asia/Manila',
      'ICT': 'Asia/Bangkok',
      'TRT': 'Europe/Istanbul',
    };
    return abbrevMap[tz] ?? LocationData.fallback.timezone;
  }

  Future<void> _persist(LocationData d) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setDouble(_kLat, d.lat);
    await prefs.setDouble(_kLon, d.lon);
    await prefs.setString(_kCity, d.city);
    await prefs.setString(_kCountry, d.country);
    await prefs.setString(_kTz, d.timezone);
    await prefs.setString(_kSource, d.source);
  }
}

/// Reactive wrapper so UI rebuilds when location changes.
class LocationNotifier extends ChangeNotifier {
  final LocationService _svc = LocationService();

  LocationData _data = LocationData.fallback;
  LocationData get data => _data;

  Future<void> load() async {
    _data = await _svc.loadSaved();
    notifyListeners();
  }

  Future<void> detect() async {
    _data = await _svc.detect();
    // Invalidate prayer cache so SalahScreen refetches with new coords
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('cached_prayer_date');
    notifyListeners();
  }
}
